File: aprendi-data/models/tables.py
"""
This module contains the AprendiTable class, which is a model for the Aprendi
"""
# from pynamodb.indexes import GlobalSecondaryIndex, AllProjection
import os
from pynamodb.models import Model
from pynamodb.attributes import UnicodeAttribute
from pynamodb.indexes import LocalSecondaryIndex, AllProjection

env = os.environ.get('ENV', 'dev')


class OrganizationTable(Model):
    """
    This class is a model for the Organization table
    """
    class Meta:
        """
        This class is a meta class for the Organization table
        """
        table_name = "aprendi_organization_table"
        region = 'us-west-2'
        host = os.environ.get('DYNAMODB_HOST', 'https://dynamodb.us-west-2.amazonaws.com')

    pk = UnicodeAttribute(hash_key=True)
    sk = UnicodeAttribute(range_key=True)
    name = UnicodeAttribute(null=False)
    data = UnicodeAttribute(null=True)


class LocalSecondaryIndex1(LocalSecondaryIndex):
    """
    This class represents a GSI for querying by teacher_id and period.
    """
    class Meta:
        """
        This class is a meta class for LocalSecondaryIndex1
        """
        projection = AllProjection()

    # These attributes are for the GSI's hash and range key
    pk = UnicodeAttribute(hash_key=True)
    lsi_sk1 = UnicodeAttribute(range_key=True)


class LocalSecondaryIndex2(LocalSecondaryIndex):
    """
    This class represents a GSI for querying by teacher_id and period.
    """
    class Meta:
        """
        This class is a meta class for LocalSecondaryIndex2
        """
        projection = AllProjection()

    # These attributes are for the GSI's hash and range key
    pk = UnicodeAttribute(hash_key=True)
    lsi_sk2 = UnicodeAttribute(range_key=True)


class OrganizationDataTable(Model):
    """
    This class is a model for the Aprendi table
    """
    class Meta:
        """
        This class is a meta class for the Aprendi table
        """
        table_name = "aprendi_organization_data_table"
        index = 'teacher_period_index'
        region = 'us-west-2'
        host = os.environ.get('DYNAMODB_HOST', 'https://dynamodb.us-west-2.amazonaws.com')

    # table keys
    pk = UnicodeAttribute(hash_key=True)
    sk = UnicodeAttribute(range_key=True)

    # student, teacher
    first_name = UnicodeAttribute(null=True)
    last_name = UnicodeAttribute(null=True)

    # course
    course_name = UnicodeAttribute(null=True)
    course_description = UnicodeAttribute(null=True)
    course_section = UnicodeAttribute(null=True)

    # course schedule
    term_name = UnicodeAttribute(null=True)

    # other data
    data = UnicodeAttribute(null=True)

    # local secondary index 1
    local_secondary_index1 = LocalSecondaryIndex1()
    lsi_sk1 = UnicodeAttribute(null=True)

    # local secondary index 2
    local_secondary_index2 = LocalSecondaryIndex2()
    lsi_sk2 = UnicodeAttribute(null=True)

File: aprendi-api/cmd/aprendi-api/models/tables.go
// Package models contains the models used in the Aprendi application
package models

// Organization represents the structure of the "aprendi_organization_table" in DynamoDB
type Organization struct {
	PK   string  `dynamo:"pk,hash"`        // Primary Key
	SK   string  `dynamo:"sk,range"`       // Sort Key
	Name string  `dynamo:"name"`           // Organization's name
	Data *string `dynamo:"data,omitempty"` // Optional data
}

// OrganizationData represents the structure of the "aprendi_organization_data_table" in DynamoDB
type OrganizationData struct {
	PK                string  `dynamo:"pk,hash"`                      // Primary Key
	SK                string  `dynamo:"sk,range"`                     // Sort Key
	FirstName         *string `dynamo:"first_name,omitempty"`         // Student/Teacher's first name
	LastName          *string `dynamo:"last_name,omitempty"`          // Student/Teacher's last name
	CourseName        *string `dynamo:"course_name,omitempty"`        // Course name
	CourseDescription *string `dynamo:"course_description,omitempty"` // Course description
	CourseSection     *string `dynamo:"course_section,omitempty"`     // Course section
	TermName          *string `dynamo:"term_name,omitempty"`          // Term name
	Data              *string `dynamo:"data,omitempty"`               // Other optional data
	LSISK1            *string `dynamo:"lsi_sk1,omitempty"`            // Local Secondary Index 1
	LSISK2            *string `dynamo:"lsi_sk2,omitempty"`            // Local Secondary Index 2
}
